lib/puppet/ssl/host.rb

```diff
@@ -192,14 +195,108 @@ def generate_certificate_request(options = {})
     true
   end
 
+  # Loads the certificate for this host. If configured to use an
+  # in-memory indirector terminus, it attempts to retrieve the cert with
+  # `indirection.find`. For all other cases, it first attempts to load
+  # the cert from disk, then to download it from the CA if this fails.
+  #
+  # The indirectory memory terminus is only used for testing the Ruby CA,
+  # so when that gets removed, this logic can go with it.
+  #
+  # @param [String] cert_name the name of the cert to load
+  # @return Puppet::SSL::Certificate if found, nil otherwise
+  def get_cert(cert_name)
+    # The memory terminus is used for testing
+    if Puppet::SSL::Certificate.indirection.terminus_class == :memory
+      return Puppet::SSL::Certificate.indirection.find(cert_name)
+    end
+
+    file_path = cert_location(cert_name)
+    if Puppet::FileSystem.exist?(file_path)
+      return Puppet::SSL::Certificate.from_s(Puppet::FileSystem.read(file_path))
+    end
+
+    if Puppet::SSL::Host.ca_location == :remote
+      # Cert not found on disk, and we have a remote ca,
+      # so attempt to download it and save it to file_path
+      download_and_save_cert(cert_name, file_path)
+    end
```

Comment:
This and the above `if` block can be combined into an `if/else` for easier verification.